From 03c599c9d2d5a42b4aea6db66e3a50263fb78440 Mon Sep 17 00:00:00 2001
From: Govindarajan Soundararajan <govindarajan.s@nutanix.com>
Date: Thu, 30 Jan 2025 02:01:28 -0800
Subject: [PATCH] fence_nutanix_ahv: Handle API rate limits (#604)

Nutanix Prism Central has rate limits set on
various API end points. The fence agent should
handle such rate limit errors and retry API requests
appropriately if it encounters any such errors. This
change adds retries for HTTP error 429 (rate limit),
and also for HTTP errors 500, and 503.
---
 agents/nutanix_ahv/fence_nutanix_ahv.py | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/agents/nutanix_ahv/fence_nutanix_ahv.py b/agents/nutanix_ahv/fence_nutanix_ahv.py
index 1a422ddad..67e6d907c 100644
--- a/agents/nutanix_ahv/fence_nutanix_ahv.py
+++ b/agents/nutanix_ahv/fence_nutanix_ahv.py
@@ -10,6 +10,8 @@
 import time
 import uuid
 import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
 
 sys.path.append("@FENCEAGENTSLIBDIR@")
 from fencing import *
@@ -20,6 +22,7 @@
 MIN_TIMEOUT = 60
 PC_PORT = 9440
 POWER_STATES = {"ON": "on", "OFF": "off", "PAUSED": "off", "UNKNOWN": "unknown"}
+MAX_RETRIES = 5
 
 
 class NutanixClientException(Exception):
@@ -44,22 +47,28 @@ def __init__(self, username, password, disable_warnings=False):
         self.password = password
         self.valid_status_codes = [200, 202]
         self.disable_warnings = disable_warnings
+        self.session = requests.Session()
+        self.session.auth = (self.username, self.password)
+
+        retry_strategy = Retry(total=MAX_RETRIES,
+                               backoff_factor=1,
+                               status_forcelist=[429, 500, 503])
+
+        self.session.mount("https://", HTTPAdapter(max_retries=retry_strategy))
 
     def request(self, url, method='GET', headers=None, **kwargs):
-        session = requests.Session()
-        session.auth = (self.username, self.password)
 
         if self.disable_warnings:
             requests.packages.urllib3.disable_warnings()
 
         if headers:
-            session.headers.update(headers)
+            self.session.headers.update(headers)
 
         response = None
 
         try:
             logging.debug("Sending %s request to %s", method, url)
-            response = session.request(method, url, **kwargs)
+            response = self.session.request(method, url, **kwargs)
             response.raise_for_status()
         except requests.exceptions.SSLError as err:
             logging.error("Secure connection failed, verify SSL certificate")
